internal/gir/types/template.go

```diff
@@ -40,79 +40,100 @@ func (f *funcArgsTemplate) AddAPI(t string, n string, k Kind, ns string, nullabl
 	if strings.ToLower(ns) == "glib" {
 		glibNs = ""
 	}
-	switch k {
-	case CallbackType:
-		if nullable {
-			c = fmt.Sprintf("%sNewCallbackNullable(%s)", glibNs, n)
-		} else {
-			c = fmt.Sprintf("%sNewCallback(%s)", glibNs, n)
-		}
-		t = "*" + t
-	case ClassesType:
+
+	if isOut {
 		if stars == 0 {
-			c = n
-			t = "uintptr"
-		} else if stars > 1 {
-			c = fmt.Sprintf("%sConvertPtr(%s)", gobjectNs, n)
-		} else if stars == 1 {
-			c = n + ".GoPointer()"
+			// For out parameters, the C type already has a pointer, and so do non-primitive Go types.
+			// For primitive Go types we need to manually add the *
+			t = "*" + t
 		}
-	case InterfacesType:
-		t = strings.TrimPrefix(t, "*")
-		if stars == 0 {
-			c = n
-			t = "uintptr"
-		} else if stars > 1 {
-			c = fmt.Sprintf("%sConvertPtr(%s)", gobjectNs, n)
-		} else if stars == 1 {
-			c = n + ".GoPointer()"
+		c = n
+	} else {
+		switch k {
+		case CallbackType:
+			if nullable {
+				c = fmt.Sprintf("%sNewCallbackNullable(%s)", glibNs, n)
+			} else {
+				c = fmt.Sprintf("%sNewCallback(%s)", glibNs, n)
+			}
+			t = "*" + t
+		case ClassesType:
+			if stars == 0 {
+				c = n
+				t = "uintptr"
+			} else if stars > 1 {
+				c = fmt.Sprintf("%sConvertPtr(%s)", gobjectNs, n)
+			} else if stars == 1 {
+				c = n + ".GoPointer()"
+			}
+		case InterfacesType:
+			t = strings.TrimPrefix(t, "*")
+			if stars == 0 {
+				c = n
+				t = "uintptr"
+			} else if stars > 1 {
+				c = fmt.Sprintf("%sConvertPtr(%s)", gobjectNs, n)
+			} else if stars == 1 {
+				c = n + ".GoPointer()"
+			}
```

Comment:
maybe you can rearrange these two statements (handle InterfacesType first), set InterfacesType to 

```
t = strings.TrimPrefix(t, "*")
fallthrough
```

just an idea